File: campaign_manager/gos/serializers/spell_slot.py
from rest_framework import serializers

from gos.models.spell import Spell
from gos.models.spell_slot import SpellSlot
from gos.serializers.spell import SpellDetailSerializer


class SpellSlotSerializer(serializers.ModelSerializer):
    spells = SpellDetailSerializer(many=True)

    class Meta:
        model = SpellSlot
        fields = [
            'id',
            'created',
            'spells',
            'level',
            'max',
            'used',
        ]
    
    def create(self, validated_data):
        spells_data = validated_data.pop('spells')
        spell_slot = SpellSlot.objects.create(**validated_data)
        return spell_slot

File: src/app/class/spell-slot/spell-slot.ts
import { Spell } from '../spell/spell';

export class SpellSlot {
    id:number;
    created:string;
    spells: Spell[];
    level: string;
    max: number;
    used: number;

    constructor(props:any) {
        this.id = props.id;
        this.created = props.created;
        this.spells = props.spells.map((spell: any) => new Spell(spell));
        this.level = props.level;
        this.max = props.max;
        this.used = props.used;
    }
}
